### DISCUSSION TITLE
Iterated filtering and mif2 parameter transformation

### DISCUSSION THREAD

**Discussion by Fuhan-Yang** [kingaa/pomp#189] [Q&A] [answered]
Hi Dr.King, 

I have two short questions about the iterated filtering. 

First, in lesson 4 from the short course Simulation-based inference, I noticed that iterated filtering was used to explore the local and global likelihood surface. Given the likelihood from IF is not the 'true likelihood', I'm curious why IF is still used in this case, rather than the likelihood evaluated by `pfilter()`? What is the difference in plotting likelihood surface between `mif2()` and `pfilter()`? 

Second, I would like to constrain my parameters when using IF, and `parameter_trans` makes this easier (thanks!!). But I have difficulties understanding the arguments `toEst` and `fromEst`. My specific question is that, say if I want my parameter to be within [0.01,0.4], what can I do?

Thanks for your help!! And please let me know if these two questions need to be separated and I will start a new discussion.

**Comment by kingaa** [comment, marked as answer]
These are good questions @Fuhan-Yang.  

First, the IF2 theorem tells us that, if we do sufficiently many IF2 iterations, reducing the "temperature" (i.e., the magnitude of the random parameter perturbations that IF2 applies) slowly enough, then IF2 will converge to the maximum-likelihood estimate (MLE).  As a byproduct of its computations, IF2 computes the likelihood of the perturbed model.  As we are at pains to explain in the [lesson](https://kingaa.github.io/pomp/FAQ.html#mif-vs-pfilter-likelihood) and in [FAQ 5.1](https://kingaa.github.io/pomp/FAQ.html#mif-vs-pfilter-likelihood), this likelihood is not the same as that for the unperturbed model.  Therefore, if one wants an accurate estimate of the likelihood at the MLE (or at any other point in the parameter space), one should use the particle filter on the unperturbed model, i.e., `pfilter()`.  You ask "What is the difference in plotting likelihood surface between `mif2()` and `pfilter()`?"  Neither of these functions plots anything in itself, so I'm not sure I understand your question.

As for your second question, if you call `parameter_trans()` with the `toEst` and `fromEst` arguments, `toEst` is supposed to transform parameters *to* the estimation scale and `fromEst`, *from* the estimation scale (i.e., back to the natural scale).  For the specific case you inquire about, supposing the parameter you want to constrain is called `theta`, you might try something like
```
parameter_trans(
  toEst=Csnippet("T_theta = logit((theta-0.01)/0.39);"),
  fromEst=Csnippet("theta = 0.39*(0.01+expit(T_theta));")
)
```
Here, we use the fact that $0.4-0.01=0.39$.  In these C snippets (as documented in the [manual](https://kingaa.github.io/manuals/pomp/html/parameter_trans.html)), `theta` is the parameter on the natural scale and `T_theta` is its transformed version.

**Reply by Fuhan-Yang** [reply]
Thank you for the replies. 

For the first question, given the likelihood from IF2 is not the true likelihood, does it mean this likelihood only serves to show the traces checking if we are approaching MLE as part of diagnostics? Besides, can we use the likelihood from `pfilter()` to explore the parameter space (like traditional grid search) before running any IF2, because `pfilter()` might be faster? If this is true, why do we still use the likelihood from `mif2()` to explore the parameter space rather than `pfilter()`? What is the difference in global search between these two methods? Sorry about the confusion!

The second question is resolved, thanks for the explanation!

**Reply by kingaa** [reply]
You are correct that the likelihood computed by `mif2`, and displayed in trace plots, is mainly useful for diagnostic purposes.  That is, it gives information that can help you to recognize convergence and to tune the algorithmic parameters for faster convergence.

Of course, you can also use grid-search, or any one of many other optimization algorithms to maximize the likelihood computed by `pfilter`.  We discuss this a little bit in [the SBIED lesson on particle filtering](https://kingaa.github.io/sbied/pfilter/) and give an example in [one of the appendices to that lesson](https://kingaa.github.io/sbied/pfilter/pf-in-Nelder-Mead.html).  Typically, a grid search will be extremely inefficient.  IF2 is very fast at finding the MLE, which is why it is recommended.    It is perhaps a little counterintuitive that, by fuzzing out the likelihood surface, one can find its maxima more easily.  To understand why it works, you can look at the [SBIED lesson on iterated filtering](https://kingaa.github.io/sbied/mif), which gives a heuristic explanation, or the [IF2 paper](https://doi.org/10.1073/pnas.1410597112) which gives a mathematical proof.

**Reply by Fuhan-Yang** [reply]
I see. My confusion is why did we use IF2 to choose a good starting value within a wide parameter box, instead of using `pfilter()`. If direct likelihood evaluation from `pfilter()` can give us a pretty good starting point, can we start by evaluating all the parameters using `pfilter()`, and then using the best parameters (parameters with the highest likelihood) found by `pfilter()` as the initial values of `mif2()`? In this way, we can save some time, because we will roughly know the parameter area that is likely to give us the global maximum, then we start to use `mif2()` from there.

**Comment by kingaa** [comment]
If I understand your question correctly, you are asking whether the initial screening of a box using some number, $M$, of `mif2` computations, each with $N$ iterations, was better than it would have been with some number, $P$, of `pfilter` evaluations.  One `mif2` iteration has roughly the same cost as one `pfilter` operation, so for the same cost as the `mif2` computations, we can evaluate the likelihood at $P=M N$ points in the box.  Now, the amount of work needed to adequately explore the space will depend on its dimension, the complexity of the likelihood surface, and the efficiency of any local searches that are applied.  If you can adequately explore the parameter space using a simple grid search with $P < M N$ points, and isolate a region on which to concentrate your efforts, then it would be better to do so.  The idea behind running $M$ local searches, each with $N>1$ amount of effort, is that this enables you to examine a much coarser grid.  That is, the idea is that with $M N$ amount of effort, you obtain a better exploration than you would with a grid search of size $P = M N$.  Whether this is in fact the case will necessarily depend on the model and the data.  In particular, various "no free lunch" theorems tell us that, for every optimization algorithm, there is at least one surface for which simple grid search is more efficient.  Whether the surface you are looking at is one such is another question, of course.

You could do some experiments: how large must you take $P$ to find the MLE?  Alternatively, how small can you take $M$?

**Comment by Fuhan-Yang** [comment]
This is interesting. If I understand correctly, because `mif2()` explores the likelihood surface in a random and initially broader way, it will do a better job screening all the possible jagged surface that is finer than the granularity of the grid search. While if the likelihood surface is smooth and unimodal, a direct grid search would be enough. It is interesting that `mif2()` is efficient given its likelihood is not true likelihood (while it should be the unbiased estimate of true likelihood..right?), maybe more N iterations will help to approximate the true MLE better. 

Thank you so much taking your time to answer all my questions. Your efforts in maintaining the discussion forum are invaluable for all the users like me.
